core/src/main/scala/producer/KafkaProducer.scala

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2017 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle
+package kafka
+
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+import cats.MonadError
+import cats.implicits._
+import freestyle.async._
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.clients.producer._
+import org.apache.kafka.common.{Metric, MetricName, PartitionInfo, TopicPartition}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+
+object producer {
+
+  final class KafkaProducerProvider[K, V] {
+
+    @free
+    trait Producer {
+
+      def isClosed: FS[Boolean]
+
+      def close(): FS[Unit]
+
+      def closeWaitingFor(timeout: Duration): FS[Unit]
+
+      def flush(): FS[Unit]
+
+      def metrics: FS[Map[MetricName, Metric]]
+
+      def partitionsFor(topic: Topic): FS[List[PartitionInfo]]
+
+      def send(record: ProducerRecord[K, V]): FS[RecordMetadata]
+
+      def sendToTopic(topic: Topic, record: (K, V)): FS[RecordMetadata]
+
+      def sendMany(records: List[ProducerRecord[K, V]]): FS[List[RecordMetadata]]
+
+      def sendManyToTopic(topic: Topic, records: List[(K, V)]): FS[List[RecordMetadata]]
+
+      def initTransactions(): FS[Unit]
+
+      def beginTransaction(): FS[Unit]
+
+      def sendOffsetsToTransaction(
+          offsets: Map[TopicPartition, OffsetAndMetadata],
+          consumerGroupId: String): FS[Unit]
+
+      def commitTransaction(): FS[Unit]
+
+      def abortTransaction(): FS[Unit]
+
+    }
+
+    class KafkaProducerHandler[M[_]](
+        implicit underlying: UnderlyingKafkaProducer[K, V],
+        ctx: AsyncContext[M],
+        ME: MonadError[M, Throwable])
+        extends Producer.Handler[M] {
+
+      private val atomicProducer: AtomicReference[KafkaProducer[K, V]] =
+        new AtomicReference(underlying.producer)
+
+      private val producerClosedState: AtomicReference[Boolean] =
```

Comment:
Yeah, I see. However, I was referring to the Boolean itself, instead of using a generic `AtomicReference`, we could use the `AtomicBoolean`.